service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/aws/s3storageFolder/ControlledAwsS3StorageFolderResource.java

```diff
@@ -0,0 +1,218 @@
+  @JsonIgnore
+  public WsmResourceFamily getResourceFamily() {
+    return WsmResourceFamily.AWS_S3_STORAGE_FOLDER;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  @JsonIgnore
+  public Optional<UniquenessCheckAttributes> getUniquenessCheckAttributes() {
+    return Optional.of(
+        new UniquenessCheckAttributes()
+            .uniquenessScope(UniquenessScope.GLOBAL)
+            .addParameter("prefix", prefix));
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void addCreateSteps(
+      CreateControlledResourceFlight flight,
+      String petSaEmail,
+      AuthenticatedUserRequest userRequest,
+      FlightBeanBag flightBeanBag) {
+    RetryRule cloudRetry = RetryRules.cloud();
+
+    flight.addStep(
+        new ValidateAwsS3StorageFolderCreateStep(this, flightBeanBag.getAwsCloudContextService()),
+        cloudRetry);
+    flight.addStep(
+        new CreateAwsS3StorageFolderStep(
+            this,
+            flightBeanBag.getAwsCloudContextService(),
+            userRequest,
+            flightBeanBag.getSamService()),
+        cloudRetry);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void addDeleteSteps(DeleteControlledResourcesFlight flight, FlightBeanBag flightBeanBag) {
+    flight.addStep(
+        new DeleteAwsS3StorageFolderStep(this, flightBeanBag.getAwsCloudContextService()),
+        RetryRules.cloud());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void addUpdateSteps(UpdateResourceFlight flight, FlightBeanBag flightBeanBag) {
+    // TODO(TERRA-315) Add support for UpdateAwsS3StorageFolder
+    throw new ApiException("addUpdateSteps NotImplemented");
+  }
+
+  public ApiAwsS3StorageFolderAttributes toApiAttributes() {
+    return new ApiAwsS3StorageFolderAttributes().bucketName(getBucketName()).prefix(getPrefix());
+  }
+
+  public ApiAwsS3StorageFolderResource toApiResource() {
+    return new ApiAwsS3StorageFolderResource()
+        .metadata(super.toApiMetadata())
+        .attributes(toApiAttributes());
+  }
+
+  @Override
+  public String attributesToJson() {
+    return DbSerDes.toJson(
+        new ControlledAwsS3StorageFolderAttributes(getBucketName(), getPrefix()));
+  }
+
+  @Override
+  public ApiResourceAttributesUnion toApiAttributesUnion() {
+    ApiResourceAttributesUnion union = new ApiResourceAttributesUnion();
+    union.awsS3StorageFolder(toApiAttributes());
+    return union;
+  }
+
+  @Override
+  public void validate() {
+    super.validate();
+    if (getResourceType() != WsmResourceType.CONTROLLED_AWS_S3_STORAGE_FOLDER
+        || getResourceFamily() != WsmResourceFamily.AWS_S3_STORAGE_FOLDER
+        || getStewardshipType() != StewardshipType.CONTROLLED) {
+      throw new InconsistentFieldsException("Expected CONTROLLED_AWS_S3_STORAGE_FOLDER");
+    }
+    if ((getBucketName() == null) || (getPrefix() == null) || (getRegion() == null)) {
+      throw new MissingRequiredFieldException(
+          "Missing required field for ControlledAwsS3StorageFolderResource.");
+    }
```

Comment:
I think we need to validate that the prefix is a top-level folder in the bucket (i.e. does not contain `/`). 
I think you could get into a weird state if you had nested controlled resources (i.e. one at `bucket/folder/` and one at `bucket/folder/subfolder`) where deleting one resource would affect other ones.

Reply:
Agree. already added as item in https://verily.atlassian.net/browse/TERRA-470